app/models/unsubscribe_key.rb

```diff
@@ -7,23 +7,102 @@ class UnsubscribeKey < ActiveRecord::Base
 
   before_create :generate_random_key
 
-  def self.create_key_for(user, type)
-    if Post === type
-      create(user_id: user.id, unsubscribe_key_type: "topic", topic_id: type.topic_id, post_id: type.id).key
-    else
-      create(user_id: user.id, unsubscribe_key_type: type).key
+  ALL_TYPE = 'all'
+  DIGEST_TYPE = 'digest'
+  TOPIC_TYPE = 'topic'
+
+  class << self
+    def create_key_for(user, type, post: nil)
+      new(user_id: user.id, unsubscribe_key_type: type).tap do |k|
+        if type == TOPIC_TYPE
+          k.topic_id = post.topic_id
+          k.post_id = post.id
+        end
+
+        k.save
+      end.key
+    end
+
+    def user_for_key(key)
+      where(key: key).first&.user
     end
   end
 
-  def self.user_for_key(key)
-    where(key: key).first.try(:user)
+  def prepare_unsubscribe_options(email_controller)
+    return false if unsubscribe_all_key?
+
+    email_controller.public_send("set_unsubscribe_#{unsubscribe_key_type}_options", self)
+  end
```

Comment:
Are we able to move this logic to the controller instead? A model shouldn't contain any controller specific concerns.